### PR TITLE
CGS GPU kernels

### PR DESCRIPTION
**Pull request by hartwiganzt**
This should be the GPU kernels for CGS and the tests.

### CODE REVIEW

File: cpu/solver/cgs_kernels.cpp [hunk 1]
```diff
@@ -50,8 +50,6 @@ void initialize(std::shared_ptr<const DefaultExecutor> exec,
                 matrix::Dense<ValueType> *q, matrix::Dense<ValueType> *u,
                 matrix::Dense<ValueType> *u_hat,
                 matrix::Dense<ValueType> *v_hat, matrix::Dense<ValueType> *t,
-                matrix::Dense<ValueType> *alpha, matrix::Dense<ValueType> *beta,
-                matrix::Dense<ValueType> *gamma,
```
Comment (tcojean) [inline cpu/solver/cgs_kernels.cpp]: I think you have some problem here. @gflegar put these back in 62ef76ce. It looks like you have merge/rebased with develop but did not favor the changes from develop.

File: examples/preconditioned_solver/CMakeLists.txt [hunk 2]
```diff
@@ -1,6 +0,0 @@
-add_executable(preconditioned_solver preconditioned_solver.cpp)
```
Comment (tcojean) [inline examples/preconditioned_solver/CMakeLists.txt]: This also undoes @gflegar 's commit cf02ca2f.

Comment (tcojean) [reply examples/preconditioned_solver/CMakeLists.txt]: There are much more like this, such as all the `.mtx` files.

**Comment by tcojean** [comment]
@hartwiganzt please check if I did it correctly.  One simple way to get the new changes is to:
```
git fetch
git checkout cgs_gpu_kernels
git reset --hard origin/cgs_gpu_kernels
```

This will fetch the new changes, go to the branch if you're not there and drop your local version to prefere the one I commited.

It looks like the problem was mostly due to the merge commit actually, because you merged a commit from 3 weeks ago where you had implemented some GPU kernels implemented with develop and this did not go well. A good way to do it is to use `git cherry-pick` when you want to pick up a single commit from long ago without too much trouble into where you currently are. See ` git help cherry-pick`.

**Comment by hartwiganzt** [comment]
Looks good to me. Great thanks @tcojean !

File: gpu/test/solver/cgs_kernels.cpp [hunk 3]
```diff
@@ -73,7 +81,7 @@ class Cgs : public ::testing::Test {
         return gko::test::generate_random_matrix<Mtx>(
             ref, num_rows, num_cols,
             std::uniform_int_distribution<>(num_cols, num_cols),
-            std::normal_distribution<>(-1.0, 1.0), rand_engine);
+            std::normal_distribution<>(0.0, 1.0), rand_engine);
```
Comment (tcojean) [inline gpu/test/solver/cgs_kernels.cpp]: Just to make sure (I have no idea), is this correct? The matrix should be positive?

File: gpu/test/solver/cgs_kernels.cpp [hunk 4]
```diff
@@ -131,97 +148,123 @@ class Cgs : public ::testing::Test {
...
+    gko::kernels::reference::cgs::step_3(ref, t.get(), u_hat.get(), r.get(),
+                                         x.get(), alpha.get());
+    gko::kernels::gpu::cgs::step_3(gpu, d_t.get(), d_u_hat.get(), d_r.get(),
+                                   d_x.get(), d_alpha.get());
 
     ASSERT_MTX_NEAR(d_x, x, 1e-14);
+    ASSERT_MTX_NEAR(d_r, r, 1e-14);
 }
-*/
+
```
Comment (tcojean) [inline gpu/test/solver/cgs_kernels.cpp]: I would like to see a test for apply if possible. i.e. similarly to what cg does.

File: gpu/test/solver/cgs_kernels.cpp (hunk 3 above)
Comment (hartwiganzt) [reply gpu/test/solver/cgs_kernels.cpp]: I think both should be fine.

File: gpu/test/solver/cgs_kernels.cpp (hunk 4 above)
Comment (hartwiganzt) [reply gpu/test/solver/cgs_kernels.cpp]: I did add a test like suggested.

File: gpu/test/solver/cgs_kernels.cpp (hunk 3 above)
Comment (tcojean) [reply gpu/test/solver/cgs_kernels.cpp]: In this case maybe it's better to take the more general version and allow negative numbers.

Comment (hartwiganzt) [reply gpu/test/solver/cgs_kernels.cpp]: done.

**Comment by tcojean** [comment]
Thank you for the changes. It works correctly and I don't see any problem with the code so I'll approve this.

**Pull request merged**